pkg/kubelet/kubelet.go

```diff
@@ -2813,6 +2824,11 @@ func (kl *Kubelet) setNodeStatusMachineInfo(node *api.Node) {
 				"Node %s has been rebooted, boot id: %s", kl.nodeName, info.BootID)
 		}
 		node.Status.NodeInfo.BootID = info.BootID
+
+		// If no cloud provider is defined - use the one detected by cadvisor
+		if kl.cloud == nil {
+			kl.updateCloudProviderFromMachineInfo(node, info)
```

Comment:
OK, I didn't feel too good about that either. I had those two reasons for placing it there:
* node.Spec is updated almost exclusively from ```Status``` methods.
* I didn't want to call GetCachedMachineInfo once more as it accesses those API's and that takes time.

Also ```setNodeStatus``` is called from ```initialNodeStatus``` so node.Spec will be updated from there. I now understand that node.Spec is not updated through ```tryUpdateNodeStatus``` 
so I see the problem here.
I just wanted to raise those points, @deads2k Please tell me if you still wish me to move it to ```initialNodeStatus```.